# PRACTISE CASE DAY 1 MONGODB WITH PYMONGO
## BY DWI SMARADAHANA INDRALOKA
***

### INSTALLING PYMONGO

In [1]:
!pip install pymongo

In [2]:
!pip install pymongo[srv]

### IMPORT LIBRARY

In [56]:
import pymongo
from pymongo import MongoClient
import pprint
import warnings
warnings.filterwarnings('ignore')

### CONNECTING TO CLUSTER

In [4]:
client = MongoClient('mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority')

### SHOW DATABASE NAME LIST

In [5]:
client.list_database_names()

['sample_mflix', 'admin', 'local']

### CHOOSE 'sample_mflix' AS DATABASE

In [6]:
db = client.sample_mflix

### SHOW COLLECTION NAME LIST

In [7]:
db.list_collection_names()

['clean_movies_dwilaras',
 'clean_movies_nurlailiis',
 'clean_movies_devita',
 'movies_initial',
 'clean_movies_frans',
 'clean_movies_tara',
 'clean_movies_tunggal',
 'clean_movies',
 'users',
 'clean_movies_nurrimah',
 'movies_testtopik',
 'movies_scratch',
 'clean_movies_topik',
 'movies',
 'clean_movies_afifah',
 'clean_movies_andreas',
 'clean_movies_indra_1']

### COLLECTION 'movies_initial'

**1. NUMBER OF DOCUMENT IN 'movies_initial'**

In [8]:
print('Total number of documents in movies_initial: ', db.movies_initial.count())

Total number of documents in movies_initial:  46014


**2. SHOW KEYS NAME IN 'movies_initial'**

In [9]:
keys_initial = [list(doc.keys()) for doc in db.movies_initial.find({})]
keys_initial = set([j for i in keys_initial for j in i])
keys_initial

{'_id',
 'awards',
 'cast',
 'country',
 'director',
 'fullplot',
 'genre',
 'imdbID',
 'imdbRating',
 'imdbVotes',
 'language',
 'lastupdated',
 'metacritic',
 'plot',
 'poster',
 'rating',
 'released',
 'runtime',
 'title',
 'type',
 'writer',
 'year'}

**3. NUMBER OF KEYS IN 'movies_initial'**

In [10]:
print('Total number of keys in movies_initial: ', len(keys_initial))

Total number of keys in movies_initial:  22


### COLLECTION 'movies'

**1. NUMBER OF DOCUMENT IN 'movies'**

In [11]:
print('Total number of documents in movies: ', db.movies.count())

Total number of documents in movies:  23541


**2. SHOW KEYS NAME IN 'movies'**

In [12]:
keys = [list(doc.keys()) for doc in db.movies.find({})]
keys = set([j for i in keys for j in i])
keys

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

**3. NUMBER OF KEYS IN 'movies'**

In [13]:
print('Total number of keys in movies: ', len(keys))

Total number of keys in movies:  22


### CHECK THE VALUES FORMAT FROM 'movies_initial' AND 'movies' FOR MAKE NEW COLLECTION

**1. MOVIES INITIAL**

In [64]:
pprint.pprint(list(db.movies_initial.find({'title':'Titanic','year':1997})))

[{'_id': ObjectId('5daf3e97ab0f2781ea561e46'),
  'awards': 'Won 11 Oscars. Another 116 wins & 63 nominations.',
  'cast': 'Leonardo DiCaprio, Kate Winslet, Billy Zane, Kathy Bates',
  'country': 'USA',
  'director': 'James Cameron',
  'fullplot': '84 years later, a 101-year-old woman named Rose DeWitt Bukater '
              'tells the story to her granddaughter Lizzy Calvert, Brock '
              'Lovett, Lewis Bodine, Bobby Buell and Anatoly Mikailavich on '
              'the Keldysh about her life set in April 10th 1912, on a ship '
              'called Titanic when young Rose boards the departing ship with '
              'the upper-class passengers and her mother, Ruth DeWitt Bukater, '
              'and her fianc�, Caledon Hockley. Meanwhile, a drifter and '
              'artist named Jack Dawson and his best friend Fabrizio De Rossi '
              'win third-class tickets to the ship in a game. And she explains '
              'the whole story from departure until the deat

**2. MOVIES**

In [65]:
pprint.pprint(list(db.movies.find({'title':'Titanic','year':1997})))

[{'_id': ObjectId('573a139af29313caabcf0d74'),
  'awards': {'nominations': 63,
             'text': 'Won 11 Oscars. Another 116 wins & 63 nominations.',
             'wins': 127},
  'cast': ['Leonardo DiCaprio', 'Kate Winslet', 'Billy Zane', 'Kathy Bates'],
  'countries': ['USA'],
  'directors': ['James Cameron'],
  'fullplot': '84 years later, a 101-year-old woman named Rose DeWitt Bukater '
              'tells the story to her granddaughter Lizzy Calvert, Brock '
              'Lovett, Lewis Bodine, Bobby Buell and Anatoly Mikailavich on '
              'the Keldysh about her life set in April 10th 1912, on a ship '
              'called Titanic when young Rose boards the departing ship with '
              'the upper-class passengers and her mother, Ruth DeWitt Bukater, '
              'and her fiancè, Caledon Hockley. Meanwhile, a drifter and '
              'artist named Jack Dawson and his best friend Fabrizio De Rossi '
              'win third-class tickets to the ship in a ga

Dengan mengetahui perbedaan format pada tiap values di atas kita dapat menentukan keys apa saja yang harus ada pada New Collection dan kita dapat mengatur values dari tiap keys yang akan kita buat pada New Collection seperti dengan melakukan spliting, conditioning, renaming dan embedding.

### MAKE LIST FOR MATCHING IN MAKE NEW COLLECTION PROCESS

In [16]:
imdb = list(db.movies.find({},{'_id':0, 'imdb':1}))
imdb_list=[]
for i in imdb :
    imdb_list.append(list(i.values())[0])

imdb_list[:5]

[{'rating': 6.3, 'votes': 162, 'id': 6517},
 {'rating': 7.8, 'votes': 4680, 'id': 8133},
 {'rating': 6.8, 'votes': 489, 'id': 12512},
 {'rating': 7.8, 'votes': 3738, 'id': 13025},
 {'rating': 7.1, 'votes': 327, 'id': 15498}]

Pada pembuatan New Collection perlu dilakukan proses matching dikarenakan jumlah document pada 'movies_initial' (46014) jauh lebih banyak dari jumlah document pada 'movies' (23541). Proses matching dilakukan dengan menggunakan keys 'imdb' pada movies yang terdiri dari imdb rating, imdb votes dan imdb id.

### MAKE NEW COLLECTION

Pada pembuatan New Collection dilakukan penentuan keys dan pengaturan values dari tiap keys sebagai berikut: 
1. Spliting, dilakukan untuk melakukan pemisahan string agar menjadi list berdasarkan tanda koma 
2. Conditioning, dilakukan untuk memberikan dua values pada keys, pertama apabila values pada keys initial berupa string kosong maka pada keys New Collection akan dianggap tidak bernilai sehingga keys tidak akan muncul, kedua jika keys initial memiliki values maka pada keys New Collection akan memiliki values yang sama
3. Renaming, dilakukan jika keys New Collection memiliki nilai yang sama pada keys initial tetapi berbeda nama
4. Embedding, dilakukan untuk menggabungkan beberapa keys initial ke delam satu keys New Collection 

In [17]:
 pipeline = [
 {
 '$project': {
    ### this code is to include the keys for a new collection on database ###    
    ### conditioning "$keys" for include, "" for exclude the keys ###
    '_id': {"$cond": [{"$eq": ["$_id", ""]}, "$nonExistinField", "$_id"]},
    'fullplot': {"$cond": [{"$eq": ["$fullplot", ""]}, "$nonExistinField", "$fullplot"]},
    'lastupdated': {"$cond": [{"$eq": ["$lastupdated", ""]}, "$nonExistinField", "$lastupdated"]},
    'metacritic': {"$cond": [{"$eq": ["$metacritic", ""]}, "$nonExistinField", "$metacritic"]},
    'plot': {"$cond": [{"$eq": ["$plot", ""]}, "$nonExistinField", "$plot"]},
    'poster': {"$cond": [{"$eq": ["$poster", ""]}, "$nonExistinField", "$poster"]},
    'title': {"$cond": [{"$eq": ["$title", ""]}, "$nonExistinField", "$title"]},
    'type': {"$cond": [{"$eq": ["$type", ""]}, "$nonExistinField", "$type"]},
    'year': {"$cond": [{"$eq": ["$year", ""]}, "$nonExistinField", "$year"]},
    ### conditioning with renaming the keys ###
    'rated': {"$cond": [{"$eq": ["$rating", ""]}, "$nonExistinField", "$rating"]},
    ### here we fill 'num_mflix_comments' with empty string because we can't find the source ###
    'num_mflix_comments': '',
    ### here we split some keys into an array value, renaming as well, with conditioning###
    'cast': {"$cond": [{"$eq": ["$cast", ""]}, "$nonExistinField", {'$split': ["$cast", ", "]}]},
    'countries': {"$cond": [{"$eq": ["$country", ""]}, "$nonExistinField", {'$split': ["$country", ", "]}]},
    'directors': {"$cond": [{"$eq": ["director", ""]}, "$nonExistinField", {'$split': ["$director", ", "]}]},
    'genres': {"$cond": [{"$eq": ["$genre", ""]}, "$nonExistinField", {'$split': ["$genre", ", "]}]},
    'languages': {"$cond": [{"$eq": ["$language", ""]}, "$nonExistinField", {'$split': ["$language", ", "]}]},
    'writers': {"$cond": [{"$eq": ["$writer", ""]}, "$nonExistinField", {'$split': ["$writer", ", "]}]},
    ### adding embedded documents ###
    ### grouping the several keys into embedded documents by one keys ### 
    'imdb':  {
              'id': "$imdbID",
              'rating': "$imdbRating",
              'votes': "$imdbVotes"
             },
    'tomatoes': {
                'lastUpdated': "$lastupdated",
                'viewer': {'meter': 1, 'numReviews': 1, 'rating': "$rating"},
                'dvd': 1,
                'fresh': 1,
                'rotten': 1,
                'production': 1
                },
    'awards': {
              'nominations': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " nomination" ]}, -2]},' ']},-1]},
              'text': "$awards",
              'wins': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " win" ]}, -2]},' ']},-1]}
              },
    'runtime': {"$cond": [{'$eq': ["$runtime", ""]}, "$nonExistinField", {'$arrayElemAt':[{'$split': [ "$runtime", " min" ]}, 0]}]},
    'released': { 
                '$concat': [ "datetime.datetime(", 
                             {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 0]}, ", ",
                             {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 1]}, ", ",
                             {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 2]}, ", 0, 0)" 
                           ] 
                }
             }
 },
 {
 '$match': { "imdb": { '$in': [dict(sorted(imdb_list[x].items(), key = lambda x: x[0])) for x in range(len(imdb_list))]}}
 },
 {
 '$out': "clean_movies_indra_1"
 }
            ]

In [19]:
db.movies_initial.aggregate(pipeline)

### VALIDATION FOR NEW COLLECTION

### 1. CHECK THE SIMILARITY OF DOCUMENTS

In [71]:
database_clean = db.clean_movies_indra_1.find({}).sort('title')
database_movies = db.movies.find({}).sort('title')
if database_clean == database_movies:
    print('All documents from two collections are already same')
else:
    print('All documents from two collection are not already same')

All documents from two collection are not already same


Pada hasil di atas didapat bahwa document pada New Collection dan document pada 'movies' masih belum sama, dimana penjelasan kenapa masih tidak sama akan dilakukan pada tahapan selanjutnya.

### 2. CHECK THE NUMBER OF DOCUMENT 'clean_movies_indra_1' VS 'movies'

**2.1 COUNT THE NUMBER OF DOCUMENT**

In [30]:
print('Total number of documents in clean_movies_indra_1: ', db.clean_movies_indra_1.count())

print('Total number of documents in movies: ', db.movies.count())

if db.movies.count()==db.clean_movies_firli.count():
    print ('Number of documents from two collections are same')
else :
    print('Number of documents from two collections are not same')

Total number of documents in clean_movies_indra_1:  23539
Total number of documents in movies:  23541
Number of documents from two collections are not same


**2.2 FIND THE DOCUMENT THAT MAKES THE NUMBER OF DOCUMENTS DIFFERENT**

In [38]:
group = list(db.clean_movies_indra_1.aggregate([{"$sortByCount":"$imdb"}]))
group2 = list(db.movies.aggregate([{"$sortByCount":"$imdb"}]))
for i in group2:
  if i not in group:
    print(i)

{'_id': {'id': 5, 'rating': 6.2, 'votes': 1189}, 'count': 2}


Berdasarkan hasil di atas, didapatkan perbedaan ada pada document dengan imdb id: 5, dimana terdapat perbedaan sebanyak 2 document, untuk detail document dengan imdb id: 5 akan ditunjukkan pada proses selanjutnya.

**2.2.1 FIND DOCUMENT WITH 'imdbID': 5 FROM 'movies_initial'**

In [69]:
pprint.pprint(list(db.movies_initial.find({'imdbID':5})))

[{'_id': ObjectId('5daf3e85ab0f2781ea55d596'),
  'awards': '1 win.',
  'cast': 'Charles Kayser, John Ott',
  'country': 'USA',
  'director': 'William K.L. Dickson',
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genre': 'Short',
  'imdbID': 5,
  'imdbRating': 6.2,
  'imdbVotes': 1189,
  'language': '',
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'metacritic': '',
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'poster': '',
  'rating': 'UNRATED',
  'released': '1893-05-09',
  '

**2.2.2 FIND DOCUMENT WITH 'imdbID': 5 FROM 'movies'**

In [70]:
pprint.pprint(list(db.movies.find({'imdb.id':{'$in':[5]}})))

[{'_id': ObjectId('573a1390f29313caabcd4135'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released'

Pada detail di atas dapat kita lihat jumlah document pada 'movies_initial' dengan imdb id: 5 ada sebanyak 1 document, sedangkan jumlah document pada 'movies' dengan imdb id: 5 ada sebanyak 3 document (duplikasi), hal tersebutlah yang membuat adanya perbedaan document pada validasi nomor 1 dan perbedaan jumlah dokumen pada validasi nomor 2 ini.

### 3 CHECK WHETHER ALL FIELDS (KEYS) IN NEW COLLECTION ARE IN 'movies'

**3.1 SHOW KEYS NAME AND KEYS NUMBER IN NEW COLLECTION**

In [25]:
keys_clean = [list(doc.keys()) for doc in db.clean_movies_indra_1.find({})]
keys_clean = set([j for i in keys_clean for j in i])
keys_clean

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [26]:
print('Total number of keys in movies_initial: ', len(keys_clean))

Total number of keys in movies_initial:  22


**3.2 SHOW KEYS NAME AND KEYS NUMBER IN 'movies'**

In [27]:
keys = [list(doc.keys()) for doc in db.movies.find({})]
keys = set([j for i in keys for j in i])
keys

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [28]:
print('Total number of keys in movies_initial: ', len(keys))

Total number of keys in movies_initial:  22


**3.3 CHECK THE SIMILARITY**

In [33]:
if keys == keys_clean:
    print('Fields on two collections are same')
else:
    print('Fields on two collections are not same')

Fields on two collections are same


Berdasarkan hasil di atas, didapatkan bahwa semua fields (keys) pada New Collection ada pada 'movies'. Hal tersebut dapat dibuktikan dengan melihat kesamaan nama keys pada masing-masing collection dan kesamaan jumlah keys.

### 4. CHECK THE SIMILARITY OF VALUES AND ITS ORDER

Pada proses validasi bahwa semua values dari New collection sama dengan semua values pada 'movies' membutuhkan waktu running yang lama jika menggunakan code, sehingga pada validasi ini akan dilakukan pembuktian dengan menggunakan **pembuktian melalui kontradiksi**. Selain itu, dari awal kita sudah mengetahui bahwa terdapat keys pada New Collection yang tidak bisa kita dapatkan nilainya dari keys pada 'movies_initial'.

**4.1 SAMPLING ONE DOCUMENT FROM NEW COLLECTION**

In [83]:
pprint.pprint(db.clean_movies_indra_1.find_one())

{'_id': ObjectId('5daf3e85ab0f2781ea55d596'),
 'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
 'cast': ['Charles Kayser', 'John Ott'],
 'countries': ['USA'],
 'directors': ['William K.L. Dickson'],
 'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
             'behind it and one on either side. The smith in the middle draws '
             'a heated metal rod from the fire, places it on the anvil, and '
             'all three begin a rhythmic hammering. After several blows, the '
             'metal goes back in the fire. One smith pulls out a bottle of '
             'beer, and they each take a swig. Then, out comes the glowing '
             'metal and the hammering resumes.',
 'genres': ['Short'],
 'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
 'lastupdated': '2015-08-26 00:03:50.133000000',
 'num_mflix_comments': '',
 'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
 'rated': 'UNRATED',
 'released': 'datetime.d

**4.2 CHOOSE DOCUMENT FROM 'movies' WITH SAME TITLE**

In [78]:
pprint.pprint(list(db.movies.find({'title':'Blacksmith Scene'})))

[{'_id': ObjectId('573a1390f29313caabcd4135'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released'

Berdasarkan hasil di atas, dapat kita lihat bahwa values dari 'num_mflix_comments' pada New Collection berupa string kosong sedangkan values 'num_mflix_comments' pada 'movies' adalah 1, hal tersebut dikarenakan tidak adanya keys pada 'movies_initial' yang merepresentasikan 'num_mflix_comments'. Selain itu terdapat pula beberapa keys dan subkeys lainnya yang memiliki kasus serupa seperti subkeys pada 'tomatoes' yaitu 'meter' dan 'numReviews'. Sehingga berdasarkan hasil tersebut maka terdapat **kontradiksi** untuk semua values dari New collection sama dengan semua values pada 'movies', dengan kata lain values dari New Collection masih tidak sama dengan values pada 'movies' tetapi sebagian besar sudah sama.

## CONCLUSION

Pada proses pembuatan New Collection berdasarkan suatu collection existing dengan output yang sama dengan collection lainnya sangat membutuhkan kelengkapan data dan pengetahuan atau pemahaman yang kuat tentang collection tersebut agar hasil kita dapat sesuai dengan yang diharapkan. Pada proses yang dilakukan di atas masih terdapat validasi-validasi yang tidak berhasil divalidasi seperti kesamaan dokumen, banyaknya dokumen dan kesamaan values dimana hal tersebut terjadi karena adanya duplikasi dokumen dan tidak adanya keys pada 'movies_initial' yang dapat merepresentasikan keys pada New Collection.

***
## THANK YOU